### PR TITLE
Add initial basic Sphinx documentation 

### PR DESCRIPTION
**Pull request by A-Joshi**
Add the basic Sphinx setup to create documentation for the mentat module. This is done by

- Creating the docs/source directory, adding in the mininmal required sphinx code (using sphinx-quickstart and sphinx-apidocs).
- Change the theme to sphinx_rtd_theme.

To generate the documents change to the docs folder and run the commane
`make clean html`

The resulting dpcumentation will be in the docs/build directory and can be viewed using your browser e.g. `firefox docs/build/html/index.html`

### CODE REVIEW

**Comment by A-Joshi** [comment]
closing to force recheck

**Comment by jakethekoenig** [comment]
Thanks so much for doing this! When I run `make clean html` I get the following error:
```
Removing everything under 'build'...
Running Sphinx v7.2.6
making output directory... done
WARNING: html_static_path entry '_static' does not exist

Theme error:
no theme named 'sphinx_rtd_theme' found (missing theme.conf?)
make: *** [Makefile:20: html] Error 2
```
Does `_static` need to be added?

**Comment by jakethekoenig** [comment]
If you run `black --preview docs` from root it will reformat your conf to satisfy our linter.

**Comment by A-Joshi** [comment]
> Thanks so much for doing this! When I run `make clean html` I get the following error:
> 
> ```
> Removing everything under 'build'...
> Running Sphinx v7.2.6
> making output directory... done
> WARNING: html_static_path entry '_static' does not exist
> 
> Theme error:
> no theme named 'sphinx_rtd_theme' found (missing theme.conf?)
> make: *** [Makefile:20: html] Error 2
> ```
> 
> Does `_static` need to be added?

Not at this time its an empty directory for now - as we add in more docs  / use other sphinx stuff and add it in the directory it will get used and come up in git.

**Comment by jakethekoenig** [comment]
> > Thanks so much for doing this! When I run `make clean html` I get the following error:
> > ```
> > Removing everything under 'build'...
> > Running Sphinx v7.2.6
> > making output directory... done
> > WARNING: html_static_path entry '_static' does not exist
> > 
> > Theme error:
> > no theme named 'sphinx_rtd_theme' found (missing theme.conf?)
> > make: *** [Makefile:20: html] Error 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does `_static` need to be added?
> 
> Not at this time its an empty directory for now - as we add in more docs / use other sphinx stuff and add it in the directory it will get used and come up in git.

Ah I forgot to install requirements. Works great.

**Comment by A-Joshi** [comment]
> > > Thanks so much for doing this! When I run `make clean html` I get the following error:
> > > ```
> > > Removing everything under 'build'...
> > > Running Sphinx v7.2.6
> > > making output directory... done
> > > WARNING: html_static_path entry '_static' does not exist
> > > 
> > > Theme error:
> > > no theme named 'sphinx_rtd_theme' found (missing theme.conf?)
> > > make: *** [Makefile:20: html] Error 2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Does `_static` need to be added?
> > 
> > 
> > Not at this time its an empty directory for now - as we add in more docs / use other sphinx stuff and add it in the directory it will get used and come up in git.
> 
> Ah I forgot to install requirements. Works great.

Not sure about a couple of things - in reality sphinx should be in the dev-requirements and not requirements (strictly speaking mentat does not require sphinx). Also will need to figure out how to build and deploy the documentation to the readthedocs.io or some location so the documentation is also automatically built and released.

**Comment by jakethekoenig** [comment]
> Not sure about a couple of things - in reality sphinx should be in the dev-requirements and not requirements (strictly speaking mentat does not require sphinx). Also will need to figure out how to build and deploy the documentation to the readthedocs.io or some location so the documentation is also automatically built and released.

I agree the sphinx requirements should be in `dev-requirements.txt`. I don't think we need to figure out hosting in this PR but I have two thoughts:
* Maybe we should check the contents of `_build` into version control and make a `/docs` command that opens them in the user's browser. Could be kind of cool.
* I looked at readthedocs a little bit and it seemed like a little overkill. I was thinking a github action that builds the docs and syncs them to a public S3 bucket might be sufficient. We could make docs.mentat.ai point there.

File: docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+import os
+import sys
+
+# -- Project information -----------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+
+project = "Mentat"
+copyright = "2023, Multiple Authors"
```
Comment (jakethekoenig) [inline docs/source/conf.py:13]: Can you change this to "2023 Abante AI"?

**Review by jakethekoenig — COMMENTED** [review]
Looks good to me as a start except for the one comment. I'd like to divide it into User Documentation and Development Documentation with the generated documentation right now being Development Documentation. But that can be done in a later PR.

**Comment by A-Joshi** [comment]
> > Not sure about a couple of things - in reality sphinx should be in the dev-requirements and not requirements (strictly speaking mentat does not require sphinx). Also will need to figure out how to build and deploy the documentation to the readthedocs.io or some location so the documentation is also automatically built and released.
> 
> I agree the sphinx requirements should be in `dev-requirements.txt`. I don't think we need to figure out hosting in this PR but I have two thoughts:
> 
>     * Maybe we should check the contents of `_build` into version control and make a `/docs` command that opens them in the user's browser. Could be kind of cool.
> 
>     * I looked at readthedocs a little bit and it seemed like a little overkill. I was thinking a github action that builds the docs and syncs them to a public S3 bucket might be sufficient. We could make docs.mentat.ai point there.

I'll move the sphinx to the dev-requirements. The documentation will eventually get complex - I see the project complexity growing and with it the documentation.


> Looks good to me as a start except for the one comment. I'd like to divide it into User Documentation and Development Documentation with the generated documentation right now being Development Documentation. But that can be done in a later PR.

Agreed.

**Comment by granawkins** [comment]
Thanks for doing this @A-Joshi! I've used readthedocs a couple times before, and what we've done is:
- Actually create a third `requirements-docs.txt`, because building a .pdf requires LaTeX which is huge. This is optional, as building the .html is actually reasonable.
- `_build` is typically temporary, not checked in (not that it can't be)
- Uploading to `readthedocs` is very simple - either to do manually or to setup a workflow. I think it's worth aiming for this.

Whatever we decide it'll be great to have this.

**Review by jakethekoenig — APPROVED** [review]

**Pull request merged**